Socket closing improvements and tests adjustments

michalpasztamobica opened the pull request:
### Description

These changes are an improvement for the hard-fault issue when closing a socket. They are based on the idea described in [this](https://github.com/ARMmbed/mbed-os/pull/8337) and [this](https://github.com/ARMmbed/mbed-os/pull/8336) by @sentinelt.
The API remains unchanged. As per API description: "Referencing a returned pointer after a close() call is not allowed and leads to undefined behaviour." - See Socket.h.
I checked that all Socket-related unittests work fine and do not hard-fault. 
In a separate commit there is also a change to icetea tests, to make them behave according to documentation. I also added a new icetea test for TcpSocket::accept() method.

### Pull request type

    [x] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

michalpasztamobica committed:
TCPSocket accept refactored to close cleanly and icetea test added

Private constructor called in TCPSocket accept, when creating a new Socket.
Close() method calls moved "up" to InternetSocket.
InternetSocket::close() returns proper error code when no socket available.
Add TcpSocket::accept icetea tests.
Deleting sockets moved to teardown.

SeppoTakalo commented on TEST_APPS/testcases/netsocket/TCPSOCKET_ACCEPT.py:
```diff
@@ -0,0 +1,102 @@
+                       purpose = "Test that TCPSocket::bind(), TCPSocket::listen() and TCPSocket::accept() works",
+                       status = "released",
+                       component= ["mbed-os", "netsocket"],
+                       author = "Juha Ylinen <juha.ylinen@arm.com>",
```
Don't copy&paste the author line.. You can remove the whole line. It is not used anywhere.

SeppoTakalo commented on TEST_APPS/testcases/netsocket/TCPSOCKET_ACCEPT.py:
```diff
@@ -0,0 +1,102 @@
+                           "duts": {
+                               '*': { #requirements for all nodes
+                                    "count":2,
+                                    "type": "hardware"
```
This is still missing the application field.

SeppoTakalo commented on TEST_APPS/testcases/netsocket/TCPSOCKET_ACCEPT.py:
```diff
@@ -0,0 +1,102 @@
+        interface = interfaceUp(self, ["dut1"])
+        self.server_ip = interface["dut1"]["ip"]
+        interface = interfaceUp(self, ["dut2"])
+        self.client_ip = interface["dut2"]["ip"]
```
This is copy&paste from old test case.. use `ifup`

michalpasztamobica replied:
I am not sure how to extract the IP address of the interface from the ifup command. And we need to know the ip address to call the connect command a few lines later.
As far as I can see the IP address will only be printed on the console...

SeppoTakalo replied:
Actually, seems I was incorrect here.. This seems OK https://github.com/ARMmbed/mbed-os/blob/master/TEST_APPS/testcases/netsocket/TCPSERVER_ACCEPT.py#L52

SeppoTakalo commented on features/netsocket/InternetSocket.cpp:
```diff
@@ -56,17 +61,15 @@ nsapi_error_t InternetSocket::close()
     nsapi_error_t ret = NSAPI_ERROR_OK;
-    if (!_stack)  {
+    if (!_socket)  {
         return ret;
```
It should return `NSAPI_ERROR_NO_SOCKET`

SeppoTakalo commented on features/netsocket/TCPSocket.cpp:
```diff
@@ -22,9 +22,19 @@ TCPSocket::TCPSocket()
+    _socket = socket,
+    _stack = parent->_stack;
+    _factory_allocated = true;
+    _remote_peer = address;
```
Would it actually compile in less code, if we would use these values in initializer?

```
TCPSocket::TCPSocket(TCPSocket* parent, nsapi_socket_t socket, SocketAddress address) :
    _socket(socket),
    _stack(parent->_stack),
    _factory_allocated(true),
    _remote_peer(address)
{
    ...
```

At least it would use copy constructor for peer address.

michalpasztamobica replied:
We cannot initialise members of the base class in the constructor initialisation list of the derived class. 
We would have to add a dedicated constructor for InternetSocket and then call:
TCPSocket::TCPSocket (...) : InternetSocket(socket, parent->_stack, ...) {}
I though this was too much of interference with the API, but if you think this is acceptable I can add the constructor.

SeppoTakalo replied:
Ah, OK. You are right.. no changes then required to this code.

KariHaapalehto commented on TEST_APPS/testcases/netsocket/TCPSERVER_ACCEPT.py:
```diff
@@ -94,9 +94,9 @@ def case(self):
-        self.command("dut2", "socket " + str(self.client_socket_id) + " delete")
+        self.command("dut1", "socket " + str(server_socket_id) + " close")
+        self.command("dut1", "socket " + str(server_base_socket_id) + " close")
+        self.command("dut2", "socket " + str(self.client_socket_id) + " close")
```
Could you please move these close-commands to teardown section.
If the test case raise error, then the close-commands are never called

KariHaapalehto commented on TEST_APPS/testcases/netsocket/TCPSOCKET_ACCEPT.py:
```diff
@@ -0,0 +1,102 @@
+            raise TestStepFail("Received data doesn't match the sent data")
+
+        response = self.command("dut1", "socket " + str(self.server_base_socket_id) + " close")
+        response = self.command("dut1", "socket " + str(socket_id) + " close")
```
Could you please move these close-commands to teardown section.
If the test case raise error, then the close-commands are never called

KariHaapalehto commented on TEST_APPS/testcases/netsocket/TCPSOCKET_ECHOTEST_BURST_SHORT.py:
```diff
@@ -72,7 +72,7 @@ def case(self):
                 raise TestStepFail("Received data doesn't match the sent data")
 
-        self.command("dut1", "socket " + str(socket_id) + " delete")
+        self.command("dut1", "socket " + str(socket_id) + " close")
```
Same as above comment

0xc0170 commented:
@michalpasztamobica When you push an update (more important after review comments), please write a comment here what has changed - to notify reviewers this is ready for another round of reviews. The question is - is it now?

michalpasztamobica commented:
@SeppoTakalo , the code is ready for review if you agree with my responses to the questions you raised. What I changed:
* updated the `cmd_socket.cpp,` to handle the `TCPSocket` properly
* applied all your and Kari's remarks apart from the two I am discussing above.
* adjusted the unit test to match the correct return value,
* I also reverted the initial idea of calling `close()` everywhere instead of `delete`. It was wrong - for the sockets created by the test using `new`, we should still call `delete `- the destructor should call `close()` for us. If we do not call `delete`, but only call `close()`, then the sockets might not be deleted (unless Python's garbage collector takes care of that? - I don't know). So in the end - we should only call `close()` 1) if we want to reopen the socket later or 2) if it was returned by `TCPSocket::accept()`.

SeppoTakalo commented on TEST_APPS/device/socket_app/cmd_socket.cpp:
```diff
@@ -1109,40 +1109,56 @@ static int cmd_socket(int argc, char *argv[])
-            return handle_nsapi_error("TCPServer::listen()", static_cast<TCPServer &>(info->socket()).listen(backlog));
+        	if (info->type() == SInfo::TCP_SERVER) {
+        		return handle_nsapi_error("TCPServer::listen()", static_cast<TCPServer &>(info->socket()).listen(backlog));
+        	} else if (info->type() == SInfo::TCP_CLIENT) {
```
Why type has to be `TCP_CLIENT` or `TCP_SERVER`
the `Socket::listen()` is in the abstract API., so we could call
```
if (cmd_parameter_int(argc, argv, "listen", &backlog)) {
    return handle_nsapi_error("Socket::listen()", info->socket().listen(backlog));
```

SeppoTakalo commented on TEST_APPS/device/socket_app/cmd_socket.cpp:
```diff
@@ -1109,40 +1109,56 @@ static int cmd_socket(int argc, char *argv[])
-                       new_info->id(), addr.get_ip_address(), addr.get_port());
-        }
-        return handle_nsapi_error("TCPServer::accept()", ret);
+    	nsapi_error_t ret;
```
Wrong intend here

SeppoTakalo commented on TEST_APPS/device/socket_app/cmd_socket.cpp:
```diff
@@ -1109,40 +1109,56 @@ static int cmd_socket(int argc, char *argv[])
-        return handle_nsapi_error("TCPServer::accept()", ret);
+    	nsapi_error_t ret;
+		if (info->type() != SInfo::TCP_SERVER) {
+			TCPSocket *new_sock = static_cast<TCPSocket &>(info->socket()).accept(&ret);
```
No, not `TCPSocket`, we should be able to use any socket here.
So `Socket *new_sock = info->socket().accept(&ret); `

SeppoTakalo commented on TEST_APPS/testcases/netsocket/TCPSOCKET_ACCEPT.py:
```diff
@@ -0,0 +1,108 @@
+        interface = interfaceUp(self, ["dut2"])
+        self.client_ip = interface["dut2"]["ip"]
+        #response = self.command("dut1", "ifup")
+        #response = self.command("dut2", "ifup")
```
Remove these

SeppoTakalo reviewed (changes requested):
Small changes still required.

Also, make sure your editor is set to use 4 spaces as a indentation. Not tabs.
Some intends seems off

SeppoTakalo commented:
And.. those `cmd_socket.cpp` changes.. check that equivalent are going to be done to Cliapp as well. I might have missed those.

The `TCPServer` accept should use the old API. But the new `accept()` should not need typecasting.

michalpasztamobica commented:
@SeppoTakalo , please review
I applied the changes you mentioned, but the cliapp repo source code for the cmd_socket.cpp has diverged quite a lot from what lives in mbed-os repo. I only picked the minimal changes to get the code compile and run so we can this get through with this pull request.
All icetea tests available in mbed-os/TEST_APPS are passing fine.

0xc0170 commented:
@KariHaapalehto review the update

0xc0170 commented:
/morph build

michalpasztamobica commented:
Test failures revealed that I did not adjust them to the changes we made.
Now when close() is called, but no socket is available, an NSAPI_ERROR_NO_SOCKET is returned (NSAPI_ERROR_OK was returned before the changes)
I need to adjust all the tests which call close() without a corresponding open() to expect an error code.

michalpasztamobica commented:
I just pushed a fix agreed with @SeppoTakalo .
The problem was that we removed the stack invalidation step from close(). Actually we do want to invalidate _stack pointer when closing the socket, because a new open() call will provide a new pointer anyway.
I locally checked that all tests which failed are now passing.

0xc0170 commented:
/morph build
